File: apps/ai-engine/kernel/graph_engine/topological_sort.py
from collections import defaultdict, deque


def topological_sort(nodes: list[str], edges: list[tuple[str, str]]) -> list[str] | None:
    in_degree: dict[str, int] = defaultdict(int)
    adjacency: dict[str, list[str]] = defaultdict(list)

    for node in nodes:
        in_degree[node] = in_degree.get(node, 0)

    for src, dst in edges:
        adjacency[src].append(dst)
        in_degree[dst] += 1

    queue: deque[str] = deque(n for n in nodes if in_degree[n] == 0)
    order: list[str] = []

    while queue:
        current = queue.popleft()
        order.append(current)
        for neighbor in adjacency[current]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    return order if len(order) == len(nodes) else None


def detect_cycle(nodes: list[str], edges: list[tuple[str, str]]) -> list[str]:
    order = topological_sort(nodes, edges)
    if order is not None:
        return []
    return [n for n in nodes if n not in (order or [])]

File: packages/graph/src/algorithms/topological-sort.ts
import { type Graph, type TopologicalSortResult } from '../types/index.js';

export const topologicalSort = <T>(graph: Graph<T>): TopologicalSortResult => {
  const adjacency = new Map<string, Set<string>>();
  const inDegree = new Map<string, number>();

  for (const node of graph.nodes) {
    adjacency.set(node.id, new Set());
    inDegree.set(node.id, 0);
  }

  for (const edge of graph.edges) {
    adjacency.get(edge.from)?.add(edge.to);
    inDegree.set(edge.to, (inDegree.get(edge.to) ?? 0) + 1);
  }

  const queue: string[] = [];
  for (const [nodeId, degree] of inDegree.entries()) {
    if (degree === 0) queue.push(nodeId);
  }

  const order: string[] = [];
  while (queue.length > 0) {
    const current = queue.shift()!;
    order.push(current);

    for (const neighbor of adjacency.get(current) ?? []) {
      const newDegree = (inDegree.get(neighbor) ?? 0) - 1;
      inDegree.set(neighbor, newDegree);
      if (newDegree === 0) queue.push(neighbor);
    }
  }

  if (order.length !== graph.nodes.length) {
    const cycle = graph.nodes
      .filter(n => !order.includes(n.id))
      .map(n => n.id);
    return { success: false, cycle };
  }

  return { success: true, order };
};
